[Jicson Toh] iP

jicsontoh opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2223S1/ip/pull/92).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main `method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Level-1, Greet, Echo, Exit

Commit:
no message

Commit:
Level-2, Add,List

Commit:
Level-3, Mark as Done

Commit:
Level-4, ToDos, Events, Deadlines

Commit:
Change name

Commit:
Level-5, Handle Errors

Commit:
Level-6, Delete

Commit:
A-TextUiTesting

Commit:
Changed data structure to ArrayList and fixed some bugs

Commit:
Changed data structure to ArrayList and fixed some bugs

Commit:
Level-7 Save the tasks in the hard disk automatically

Commit:
Level-8 Teach Duke to understand dates and times

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/DeadlineTask.java
#	src/main/java/EventTask.java

Commit:
A-MoreOOP - Refactor the code

Commit:
A-Packages - Divide classes into packages

Commit:
Merge branch 'add-gradle-support'

Commit:
A-JUnit - Add JUnit Tests

Commit:
Add checkstyle.xml and suppressions.xml

Commit:
Level-9 Find a task by searching for keyword

Commit:
A-JavaDoc- Add JavaDocs comments to code

Commit:
A-CodingStandard- Indentation of switch statements.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/util/Ui.java

Commit:
A-MoreOOP - Refactor the code with Command Classes

JovonLim commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Executes the command input.
+     * @param tasks   amends task list if any.
+     * @param ui      ui to output feedback.
```
```suggestion
     * Executes the command input.
     *
     * @param tasks   amends task list if any.
     * @param ui      ui to output feedback.
```
I think there should be an empty line between the description and parameter lines for the coding standard. I noticed that only some docs have that so it would be good to rectify the rest.

jicsontoh replied:
Will take note thanks!

JovonLim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+ * @author Jicson Toh
+ */
+public class Task {
+    protected String action;
```
```suggestion
    protected String description;
```
Perhaps this would be a better name rather than action?

jicsontoh replied:
agreed, will take note thanks!

JovonLim commented on src/main/java/duke/data/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+        return this.list;
+    }
+
+    public String getTaskString(int index) {
```
```suggestion
    public String getTaskDescription(int index) {
```
Rather than String, it might be better to refer it as the task's description

jicsontoh replied:
agreed, will take note thanks!

JovonLim commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,72 @@
+     * @param action user input.
+     * @return returns the string action.
+     */
+    public Command parseCommand(String action) throws DukeException {
```
I like that it systematically checks for the single word commands then for the other commands 👍

jicsontoh replied:
Thanks!

JovonLim commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,126 @@
+     * @param removedTask task to be removed.
+     * @param size size of list after task is remove.
+     */
+    public void removeTask(String removedTask, int size) {
```
```suggestion
    public void showRemovedTask(String removedTask, int size) {
```
Since this method is used by the Ui object, perhaps it would make more sense by naming it to indicate that its printing out a message rather than actually removing a task? Similarly for the below task related methods

jicsontoh replied:
agreed, will take note thanks!

JovonLim reviewed (commented):
Overall good job on the iP 😄. Just take note of the javadocs for the coding standard as well as a few possible improvements to the naming of your methods and fields.

Commit:
Add and use checkstyle

kangqiao322 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,71 @@
+ * @author Jicson Toh
+ */
+public class AddCommand extends Command {
+    private final String action;
```
Should constant names be in lower case? Some other constant names have the same issue too.

jicsontoh replied:
hmm its not a constant but a variable to be instantiated when creating the AddCommand Object

kangqiao322 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,71 @@
+            }
+            ui.addedTask(tasks.getTaskString(tasks.getSize() - 1), tasks.getSize());
+        } catch (Exception e) {
+            throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
```
I like how you used switch cases for numerous different conditions instead of if else. The indentation is correct as well. Great job!

jicsontoh replied:
Thank you!

kangqiao322 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+     * @param keyword input keyword.
+     * @return true if contains the keyword.
+     */
+    public boolean matchKeyword(String keyword) {
```
Perhaps try to use prefix such as is, has, was etc for boolean method names.

jicsontoh replied:
Will take note thanks!

kangqiao322 reviewed (commented):
Overall, a job well done! Just need a few changes to the naming of some variables and methods along with more JavaDocs for some public methods.

kangqiao322 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,63 @@
+ * @author Jicson Toh
+ */
+public class FindCommand extends Command {
+    private final String action;
```
I think this variable name should be in upper case as well?

jicsontoh replied:
same for this!

Commit:
Level-10 Add a GUI to Duke

Commit:
Merge branch 'branch-Level-10'

Commit:
Update file path

Commit:
Use assert feature

The assert feature is used to define assumptions about the program state so that the runtime can verify them.

Implement the assert features in methods where the execution of the method have to be of certain assumption before running.

This is to allow the program to run as per design and to catch any possible bugs and terminate the program as soon as it detects.

Commit:
Improve code quality

The default file path had to be explicitly stated when creating the Duke object and this requires the user to know the file path.

Store the default file path as a named constant.

By storing it as a named constant, the user does not need to know the file path when initialising the object. The Duke class will by default create the file in the default file path if no file path is specified.

jicsontoh committed:
Merge pull request #1 from jicsontoh/branch-A-Assertions

Use assert feature

jicsontoh committed:
Merge pull request #2 from jicsontoh/branch-A-CodeQuality

Improve code quality

jicsontoh committed:
Set up GitHub Actions to set up CI

Commit:
Add javadocs to removeTask method

jicsontoh committed:
Merge pull request #3 from jicsontoh/branch-A-CI

Set up GitHub Actions to set up CI

Commit:
Remove I/O redirection tests

Commit:
Add update feature to tasks

Users are not able to rename the task that they added to the list.

Add a new update feature to allow users to change the name of the task.

This allow the users to easily update the task instead of having to delete the task and add a new task.

jicsontoh committed:
Add files via upload

jicsontoh committed:
Update README.md

jicsontoh committed:
Update README.md

jicsontoh committed:
Update README.md

Commit:
Better GUI

jicsontoh committed:
Delete Ui.png

jicsontoh committed:
Add files via upload

jicsontoh committed:
Update README.md

Commit:
Update welcome message

jicsontoh committed:
Change Image

jicsontoh committed:
Update MainWindow.java

jicsontoh committed:
Update README

jicsontoh committed:
Update Find case-sensitive

jicsontoh committed:
Update README

jicsontoh committed:
Add demo video

jicsontoh committed:
Update README

jicsontoh committed:
Update README.md
